File: solutions/javascript/stack.js
var Stack = module.exports = function () {
  this.head   = null;
  this.length = 0;
};

Stack.prototype.push = function (value) {
  var node = {
    value: value,
    next: null
  };

  if (!this.head) {
    this.head = node;
  } else {
    node.next = this.head;
    this.head = node;
  }

  return this.length += 1;
};

Stack.prototype.pop = function () {
  // If there is no head node, return `undefined`
  if (!this.head) { return; }

  var node = this.head;

  // Update the head reference and remove the next node reference from the
  // previous head.
  this.head = node.next;
  node.next = null;

  this.length -= 1;

  return node.value;
};

File: solutions/java/Stack.java
// Stack that takes an int
public class Stack {
	private Node top;

	private class Node {
		private Node next;
		private int data;
		Node(int data, Node next) {
			this.data = data;
			this.next = next;
		}
	}

	public void push(int data) {
		if (top == null) {
			Node n = new Node(data, null);
			top = n;
		} else {
			Node n = new Node(data, top);
			top = n;
		}
	}

	public int pop() throws Exception {
		if (top == null) {
			throw new Exception("nothing in stack to pop");
		}
		int dat = top.data;
		top = top.next;
		return dat;
	}

	public void printStack(){
		Node current = top;
		while (current != null) {
			System.out.println(current.data);
			current = current.next;
		}

	}

	//Test Client
	public static void main(String[] args) throws Exception{
		Stack stack = new Stack();
		stack.push(1);
		stack.push(2);
		stack.push(3);
		stack.push(4);
		stack.push(5);
		

		stack.printStack();
		System.out.println("popped this out: " + stack.pop());
		System.out.println("popped this out: " + stack.pop());
		stack.printStack();
	}
}